webrender/src/device/gfx.rs

```diff
@@ -777,119 +777,155 @@ impl<B: hal::Backend> BufferPool<B> {
-            buffer_index: 0,
+            current_buffer_index: 0,
         }
     }
 
     pub fn add<T: Copy>(
         &mut self,
         device: &B::Device,
         data: &[T],
-        memory_types: &Vec<hal::MemoryType>,
     ) {
-        let data_len = data.len();
-        let buffer_len = data_len * self.data_stride;
-        if self.offset() * self.data_stride + buffer_len > self.buffer().buffer_size {
-            let update_len = (self.buffer().buffer_size / self.data_stride - self.offset()) &! (mem::size_of::<T>() - 1);
-            if update_len > 0 {
-                self.buffer().update(
-                    device,
-                    &data[0 .. update_len],
-                    self.offset(),
-                    self.non_coherent_atom_size,
-                );
-                self.size[self.buffer_index] = update_len;
-                self.offset[self.buffer_index] += update_len;
-            }
-            let data = &data[update_len ..];
-            self.buffer_index += 1;
-            self.offset.push(0);
-            self.size.push(0);
-            if self.buffer_pool.len() < (self.buffer_index + 1) {
-                self.buffer_pool.push(
-                    Buffer::new(
-                        device,
-                        memory_types,
-                        hal::buffer::Usage::VERTEX,
-                        self.pitch_alignment_mask,
-                        MAX_INSTANCE_COUNT,
-                        self.data_stride,
+        let mut data = data;
+        while data.len() > 0 {
+            if self.current_buffer().buffer.buffer_size - self.current_buffer().offset * self.data_stride < mem::size_of::<T>() {
+                self.current_buffer_index += 1;
+                if self.buffers.len() < self.buffer_count() {
+                    self.buffers.push(
+                        InstancePoolBuffer::new(
+                            device,
+                            &self.memory_types,
+                            hal::buffer::Usage::VERTEX,
+                            self.data_stride,
+                            self.pitch_alignment_mask,
+                        )
                     )
-                )
+                }
             }
-            self.add(device, data, memory_types);
-        } else {
-            assert!(
-                self.offset() * self.data_stride + buffer_len <= self.buffer().buffer_size,
-                "offset({:?}) * data_stride({:?}) + buffer_len({:?}) <= buffer_size({:?})",
-                self.offset(), self.data_stride, buffer_len, self.buffer().buffer_size
-            );
-            self.buffer().update(
+
+            let update_size = if (self.current_buffer().offset + data.len()) * self.data_stride > self.current_buffer().buffer.buffer_size {
+                let mut update_size = self.current_buffer().buffer.buffer_size / self.data_stride - self.current_buffer().offset;
+                update_size -= update_size % mem::size_of::<T>();
```

Comment:
This doesn't look right. `upload_size` here is the number of elements in `data` vector, each of `size_of<T>` bytes. IT doesn't make sense to compute the remainder of the element count by this byte count.

Reply:
That's right. I removed this line, since `self.data_stride` must be equal to the size of `T` (we create the `InstanceBufferHandler` with this particular data stride) and the `buffer_size` of the underlying buffer also must be a multiple of `T` for the same reason.
I put an [assert](https://github.com/szeged/webrender/pull/220/files#diff-4d269bbd98df6af055b56ab19362567fR889) in the `add` function, to check if the data stride and the size of `T` is different. If that fails, we do something wrong.